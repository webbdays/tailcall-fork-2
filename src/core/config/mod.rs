pub use apollo::*;
pub use config::*;
pub use config_module::*;
pub use key_values::*;
pub use link::*;
pub use npo::QueryPath;
pub use reader_context::*;
pub use server::*;
pub use source::*;
pub use telemetry::*;
pub use upstream::*;
mod apollo;
mod config;
mod config_module;
pub mod cors;
mod from_document;
pub mod group_by;
mod headers;
mod into_document;
mod key_values;
mod link;
mod npo;
pub mod reader;
pub mod reader_context;
mod server;
mod source;
mod telemetry;
pub mod transformer;
mod upstream;
