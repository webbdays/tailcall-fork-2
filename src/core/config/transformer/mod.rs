mod ambiguous_type;
mod consolidate_url;
mod flatten_single_field;
mod improve_type_names;
mod merge_types;
mod nested_unions;
mod preset;
mod required;
mod tree_shake;
mod union_input_type;

pub use ambiguous_type::{AmbiguousType, Resolution};
pub use consolidate_url::ConsolidateURL;
pub use flatten_single_field::FlattenSingleField;
pub use improve_type_names::ImproveTypeNames;
pub use merge_types::TypeMerger;
pub use nested_unions::NestedUnions;
pub use preset::Preset;
pub use required::Required;
pub use tree_shake::TreeShake;
pub use union_input_type::UnionInputType;
