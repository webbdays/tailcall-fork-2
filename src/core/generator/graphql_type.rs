use std::fmt::Display;

use convert_case::{Case, Casing};
pub(super) static DEFAULT_SEPARATOR: &str = "__";
static PACKAGE_SEPARATOR: &str = ".";

fn normalize_name(name: &str) -> String {
    name.replace(PACKAGE_SEPARATOR, "_")
}

/// A helper to infer and build the name of a GraphQL type from raw list of
/// strings separated by a special character. It can be used to represent any
/// kind of GraphQL entity like an enum, an object type, a field or a method. It
/// can also be used to represent a package or a namespace. In unparsed form it
/// is just a list of strings.
#[derive(Debug, Clone, PartialEq)]
pub struct GraphQLType<A>(A);

/// Represents a parsed GraphQL name where the actual name, the namespace and
/// the type of the entity is known.
#[derive(Debug, Clone, PartialEq)]
pub struct Parsed {
    namespace: Namespace,
    name: String,
    entity: Entity,
}

/// Represents an unparsed GraphQL name with just a list of strings.
/// Keeping the head separated ensures that we don't have an empty list of
/// strings.
#[derive(Debug, Clone)]
pub struct Unparsed {
    /// The head of the GraphQL type name.
    head: String,

    /// The tail of the GraphQL type name.
    tail: Vec<String>,
}

/// Represents a package or a namespace for the type, a feature typically found
/// in protobuf.
#[derive(Debug, Default, Clone, PartialEq)]
struct Namespace(Vec<String>);

impl Namespace {
    fn id(&self) -> String {
        self.0.join(PACKAGE_SEPARATOR)
    }
    fn is_empty(&self) -> bool {
        self.0.is_empty()
    }

    #[cfg(test)]
    fn new<A: AsRef<str> + Clone>(s: &[A]) -> Self {
        Self(s.iter().map(|a| a.as_ref().to_string()).collect())
    }
}

impl Display for Namespace {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(self.0.join(DEFAULT_SEPARATOR).as_str())
    }
}

impl GraphQLType<Unparsed> {
    pub fn new(name: &str) -> Self {
        Self(Unparsed { head: name.to_string(), tail: Vec::new() })
    }

    pub fn extend<S: AsRef<str>>(mut self, items: &[S]) -> Self {
        for item in items {
            self.0.tail.push(item.as_ref().to_string());
        }
        self
    }

    pub fn push<S: AsRef<str>>(mut self, text: S) -> Self {
        self.0.tail.push(text.as_ref().to_string());

        self
    }

    /// Parses list of string and extracts the entity's name, its package or
    /// namespace and it's type.
    fn parse(self, entity: Entity) -> GraphQLType<Parsed> {
        let unparsed = self.0;
        let path = unparsed
            .tail
            .iter()
            // TODO: separate should be passed as an argument.
            // Parser can not assume that the separator is always a dot. For eg: in Rust itself the
            // types are separated by `::`
            .flat_map(|c| c.split(PACKAGE_SEPARATOR))
            .map(|a| a.trim().to_string())
            .filter(|a| !a.is_empty())
            .collect::<Vec<_>>();

        let name = normalize_name(&unparsed.head);

        GraphQLType(Parsed { name, namespace: Namespace(path), entity })
    }

    pub fn into_enum(self) -> GraphQLType<Parsed> {
        self.parse(Entity::Enum)
    }

    pub fn into_enum_variant(self) -> GraphQLType<Parsed> {
        self.parse(Entity::EnumVariant)
    }

    pub fn into_object_type(self) -> GraphQLType<Parsed> {
        self.parse(Entity::ObjectType)
    }

    pub fn into_method(self) -> GraphQLType<Parsed> {
        self.parse(Entity::Method)
    }

    pub fn into_field(self) -> GraphQLType<Parsed> {
        self.parse(Entity::Field)
    }
}

impl GraphQLType<Parsed> {
    pub fn id(&self) -> String {
        let namespace = &self.0.namespace;
        if !namespace.is_empty() {
            format!("{}.{}", namespace.id(), self.0.name)
        } else {
            self.0.name.clone()
        }
    }
}

/// Used to convert proto type names to GraphQL formatted names.
/// Enum to represent the type of the descriptor
#[derive(Clone, Debug, PartialEq)]
enum Entity {
    Enum,
    EnumVariant,
    ObjectType,
    Method,
    Field,
}

impl Display for GraphQLType<Parsed> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let parsed = &self.0;
        match parsed.entity {
            Entity::EnumVariant => f.write_str(parsed.name.as_str())?,
            Entity::Field => f.write_str(parsed.name.to_case(Case::Camel).as_str())?,
            Entity::Method => {
                if !parsed.namespace.is_empty() {
                    f.write_str(parsed.namespace.to_string().as_str())?;
                    f.write_str(DEFAULT_SEPARATOR)?;
                };
                f.write_str(parsed.name.as_str())?
            }
            Entity::Enum | Entity::ObjectType => {
                if !parsed.namespace.is_empty() {
                    f.write_str(parsed.namespace.to_string().as_str())?;
                    f.write_str(DEFAULT_SEPARATOR)?;
                };
                f.write_str(parsed.name.as_str())?
            }
        };
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use pretty_assertions::assert_eq;

    use super::*;

    type TestParams = (
        (
            Entity, // Entity type for example Enum, EnumVariant, ObjectType, Method, Field
            &'static [&'static str], // Namespace or package
            &'static str, // Name of the entity
        ),
        &'static str,
    );

    #[test]
    fn test_from_enum() {
        let input: Vec<TestParams> = vec![
            // Enums
            ((Entity::Enum, &[], "foo"), "foo"),
            ((Entity::Enum, &[], "Foo"), "Foo"),
            ((Entity::Enum, &["a", "b.c"], "foo_bar"), "a__b__c__foo_bar"),
            ((Entity::Enum, &[], "a.b.c.foo"), "a_b_c_foo"),
            ((Entity::Enum, &["a.b.c"], "foo"), "a__b__c__foo"),
            (
                (Entity::Enum, &["a.b.c"], "d.e.f.foo"),
                "a__b__c__d_e_f_foo",
            ),
            ((Entity::Enum, &[""], "a.b.c.foo"), "a_b_c_foo"),
            ((Entity::Enum, &[], "a_b_c_foo"), "a_b_c_foo"),
        ];

        assert_type_names(input);
    }

    #[test]
    fn test_from_enum_variant() {
        let input: Vec<TestParams> = vec![
            // Enum variants
            ((Entity::EnumVariant, &[], "foo"), "foo"),
            ((Entity::EnumVariant, &[], "FOO_VAR"), "FOO_VAR"),
            ((Entity::EnumVariant, &[], "a.b.c.foo"), "a_b_c_foo"),
            ((Entity::EnumVariant, &["a.b.c"], "foo"), "foo"),
            ((Entity::EnumVariant, &["a.b"], "d.e.foo"), "d_e_foo"),
            ((Entity::EnumVariant, &[""], "a.b.c.foo"), "a_b_c_foo"),
            ((Entity::EnumVariant, &[], "a_b_c_foo"), "a_b_c_foo"),
        ];

        assert_type_names(input);
    }

    #[test]
    fn test_from_object_type() {
        let input: Vec<TestParams> = vec![
            // Object types
            ((Entity::ObjectType, &[], "foo"), "foo"),
            (
                (Entity::ObjectType, &["a", "b.c"], "fooBar"),
                "a__b__c__fooBar",
            ),
            ((Entity::ObjectType, &[], "a.b.c.foo"), "a_b_c_foo"),
            ((Entity::ObjectType, &["a.b.c"], "foo"), "a__b__c__foo"),
            ((Entity::ObjectType, &["a.b"], "d.e.foo"), "a__b__d_e_foo"),
            ((Entity::ObjectType, &[""], "a.b.c.foo"), "a_b_c_foo"),
            ((Entity::ObjectType, &[], "a_b_c_foo"), "a_b_c_foo"),
            ((Entity::ObjectType, &[], "foo.bar.Baz"), "foo_bar_Baz"),
        ];

        assert_type_names(input);
    }

    #[test]
    fn test_from_method() {
        let input: Vec<TestParams> = vec![
            // Methods
            ((Entity::Method, &[], "foo"), "foo"),
            ((Entity::Method, &["a.b.c"], "fooBar"), "a__b__c__fooBar"),
            (
                (Entity::Method, &["a.b", "c"], "foo_bar"),
                "a__b__c__foo_bar",
            ),
            ((Entity::Method, &[], "a.b.c.foo"), "a_b_c_foo"),
            ((Entity::Method, &["a.b.c"], "foo"), "a__b__c__foo"),
            ((Entity::Method, &["a.b"], "d.e.foo"), "a__b__d_e_foo"),
            ((Entity::Method, &[""], "a.b.c.foo"), "a_b_c_foo"),
            ((Entity::Method, &[], "a_b_c_foo"), "a_b_c_foo"),
        ];

        assert_type_names(input);
    }

    #[test]
    fn test_from_field() {
        let input: Vec<TestParams> = vec![
            // Fields
            ((Entity::Field, &[], "foo"), "foo"),
            ((Entity::Field, &["a.b.c"], "fooBar"), "fooBar"),
            ((Entity::Field, &["a.b", "c"], "foo_bar"), "fooBar"),
            ((Entity::Field, &[], "a.b.c.foo"), "aBCFoo"),
            ((Entity::Field, &["a.b.c"], "foo"), "foo"),
            ((Entity::Field, &["a.b"], "d.e.foo"), "dEFoo"),
            ((Entity::Field, &[""], "a.b.c.foo"), "aBCFoo"),
            ((Entity::Field, &[], "a_bC_foo"), "aBCFoo"),
        ];

        assert_type_names(input);
    }

    #[test]
    fn test_parse_name() {
        let actual = GraphQLType::new("foo").into_enum();
        let expected = GraphQLType(Parsed {
            name: "foo".to_string(),
            namespace: Namespace::default(),
            entity: Entity::Enum,
        });
        assert_eq!(actual, expected);
    }

    #[test]
    fn test_parse_namespace() {
        let actual = GraphQLType::new("foo").push("bar").push("baz").into_enum();
        let expected = GraphQLType(Parsed {
            name: "foo".to_string(),
            namespace: Namespace::new(&["bar", "baz"]),
            entity: Entity::Enum,
        });
        assert_eq!(actual, expected);
    }

    fn assert_type_names(input: Vec<TestParams>) {
        for ((entity, namespaces, name), expected) in input {
            let mut g = GraphQLType::new(name);
            for namespace in namespaces {
                g = g.push(namespace);
            }

            let actual = g.parse(entity).to_string();
            assert_eq!(actual, expected);
        }
    }
}
