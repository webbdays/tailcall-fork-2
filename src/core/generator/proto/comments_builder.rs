use prost_reflect::prost_types::SourceCodeInfo;

pub struct CommentsBuilder {
    source_code_info: Option<SourceCodeInfo>,
}

impl CommentsBuilder {
    pub fn new(source_code_info: Option<SourceCodeInfo>) -> Self {
        Self { source_code_info }
    }

    pub fn get_comments(&self, path: &[i32]) -> Option<String> {
        self.source_code_info.as_ref().and_then(|info| {
            info.location
                .iter()
                .find(|loc| loc.path == path)
                .and_then(|loc| {
                    loc.leading_comments.as_ref().map(|c| {
                        c.lines()
                            .map(|line| line.trim_start_matches('*').trim())
                            .filter(|line| !line.is_empty())
                            .collect::<Vec<_>>()
                            .join("\n  ")
                    })
                })
        })
    }
}
