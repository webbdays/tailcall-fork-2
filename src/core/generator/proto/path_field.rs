pub enum PathField {
    EnumType,
    MessageType,
    Service,
    Field,
    Method,
    EnumValue,
    NestedType,
    NestedEnum,
}

impl PathField {
    pub fn value(&self) -> i32 {
        match self {
            PathField::EnumType => 5,
            PathField::MessageType => 4,
            PathField::Service => 6,
            PathField::Field => 2,
            PathField::Method => 2,
            PathField::EnumValue => 2,
            PathField::NestedType => 3,
            PathField::NestedEnum => 4,
        }
    }
}
