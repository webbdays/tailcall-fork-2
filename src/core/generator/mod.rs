mod from_json;
mod from_proto;
mod generator;
mod graphql_type;
mod json;
mod proto;

pub use from_json::{FromJsonGenerator, RequestSample};
pub use generator::{Generator, Input};

use crate::core::counter::{Count, Counter};

pub struct NameGenerator {
    counter: Counter<u64>,
    prefix: String,
}

impl NameGenerator {
    pub fn new(prefix: &str) -> Self {
        Self { counter: Counter::new(1), prefix: prefix.to_string() }
    }

    pub fn next(&self) -> String {
        let id = self.counter.next();
        format!("{}{}", self.prefix, id)
    }
}
