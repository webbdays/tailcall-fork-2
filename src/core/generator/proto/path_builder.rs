use super::path_field::PathField;

pub struct PathBuilder {
    base_path: Vec<i32>,
}

impl PathBuilder {
    pub fn new(base_path: &[i32]) -> Self {
        Self { base_path: base_path.to_vec() }
    }

    pub fn extend(&self, field: PathField, index: i32) -> Vec<i32> {
        let mut extended_path = self.base_path.clone();
        extended_path.push(field.value());
        extended_path.push(index);
        extended_path
    }
}
