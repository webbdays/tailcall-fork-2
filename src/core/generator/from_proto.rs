use std::collections::{BTreeSet, HashSet};

use anyhow::{bail, Result};
use derive_setters::Setters;
use prost_reflect::prost_types::field_descriptor_proto::Label;
use prost_reflect::prost_types::{
    DescriptorProto, EnumDescriptorProto, FileDescriptorSet, ServiceDescriptorProto, SourceCodeInfo,
};
use serde_json::Value;

use super::graphql_type::{GraphQLType, Unparsed};
use super::proto::comments_builder::CommentsBuilder;
use super::proto::path_builder::PathBuilder;
use super::proto::path_field::PathField;
use crate::core::config::transformer::{AmbiguousType, TreeShake};
use crate::core::config::{Arg, Config, Enum, Field, Grpc, Tag, Type, Union, Variant};
use crate::core::transform::{Transform, TransformerOps};
use crate::core::valid::Validator;

/// Assists in the mapping and retrieval of proto type names to custom formatted
/// strings based on the descriptor type.
#[derive(Setters)]
struct Context {
    /// The current proto package name.
    namespace: Vec<String>,

    /// Final configuration that's being built up.
    config: Config,

    /// Root GraphQL query type
    query: String,

    /// Set of visited map types
    map_types: HashSet<String>,

    /// Optional field to store source code information, including comments, for
    /// each entity.
    comments_builder: CommentsBuilder,
}

impl Context {
    fn new(query: &str) -> Self {
        Self {
            query: query.to_string(),
            namespace: Default::default(),
            config: Default::default(),
            map_types: Default::default(),
            comments_builder: CommentsBuilder::new(None),
        }
    }

    /// Sets source code information for preservation of comments.
    fn with_source_code_info(mut self, source_code_info: SourceCodeInfo) -> Self {
        self.comments_builder = CommentsBuilder::new(Some(source_code_info));
        self
    }

    /// Resolves the actual name and inserts the type.
    fn insert_type(mut self, name: String, ty: Type) -> Self {
        self.config.types.insert(name.to_string(), ty);
        self
    }

    /// Converts oneof definitions in message to set of types with union
    fn insert_oneofs(
        mut self,
        type_name: String, // name of the message
        base_type: Type,   // that's the type with fields that are not oneofs
        oneof_fields: Vec<Vec<(String, Field)>>, /* there is multiple oneof definitions every
                            * one of which contains multiple fields */
    ) -> Self {
        fn collect_types(
            type_name: String,
            base_type: Type,
            oneof_fields: &[Vec<(String, Field)>], // currently processed set of oneof fields
            output: &mut Vec<(String, Type)>,      // newly generated types with their names
        ) {
            let Some(one_of) = oneof_fields.first() else {
                output.push((type_name, base_type));

                return;
            };
            let oneof_fields = &oneof_fields[1..];

            // is there is only one variant for oneof field
            // this field is actually an optional field
            // for graphql and we can generate only single variant for this
            if one_of.len() == 1 {
                let (field_name, field) = one_of[0].clone();
                let mut new_type = base_type;
                new_type.fields.insert(field_name, field);

                collect_types(format!("{type_name}__Var"), new_type, oneof_fields, output);

                return;
            }

            // since all of the fields are actually optional in protobuf generate also a
            // type without this oneof completely
            collect_types(
                format!("{type_name}__Var"),
                base_type.clone(),
                oneof_fields,
                output,
            );

            for (i, (field_name, field)) in one_of.iter().enumerate() {
                let mut new_type = base_type.clone();
                let mut field = field.clone();

                // mark this field as required to force type-check on specific variant of oneof
                field.required = true;

                // add new field specific to this variant of oneof field
                new_type.fields.insert(field_name.clone(), field);

                collect_types(
                    format!("{type_name}__Var{i}"),
                    new_type,
                    oneof_fields,
                    output,
                );
            }
        }

        let mut union_types = Vec::new();

        collect_types(
            type_name.clone(),
            base_type,
            &oneof_fields,
            &mut union_types,
        );

        // if there is only one type in union no need
        // to actually create union and use just this type
        if union_types.len() == 1 {
            let (_, ty) = union_types.pop().unwrap();
            self.config.types.insert(type_name, ty);
            return self;
        }

        let mut union_ = Union::default();

        for (type_name, ty) in union_types {
            union_.types.insert(type_name.clone());

            self = self.insert_type(type_name, ty);
        }

        self.config.unions.insert(type_name, union_);

        self
    }

    /// Processes proto enum types.
    fn append_enums(
        mut self,
        enums: &[EnumDescriptorProto],
        parent_path: &PathBuilder,
        is_nested: bool,
    ) -> Self {
        for (index, enum_) in enums.iter().enumerate() {
            let enum_name = enum_.name();

            let enum_type_path = if is_nested {
                parent_path.extend(PathField::NestedEnum, index as i32)
            } else {
                parent_path.extend(PathField::EnumType, index as i32)
            };

            let mut variants_with_comments = BTreeSet::new();

            for (value_index, v) in enum_.value.iter().enumerate() {
                let variant_name = GraphQLType::new(v.name()).into_enum_variant().to_string();

                // Path to the enum value's comments
                let value_path = PathBuilder::new(&enum_type_path)
                    .extend(PathField::EnumValue, value_index as i32); // 2: value field

                // Get comments for the enum value
                let comment = self.comments_builder.get_comments(&value_path);

                // Format the variant with its comment as description
                if let Some(comment) = comment {
                    // TODO: better support for enum variant descriptions [There is no way to define
                    // description for enum variant in current config structure]
                    let variant_with_comment =
                        format!("\"\"\n  {}\n  \"\"\n  {}", comment, variant_name);
                    variants_with_comments.insert(variant_with_comment);
                } else {
                    variants_with_comments.insert(variant_name);
                }
            }

            let type_name = GraphQLType::new(enum_name)
                .extend(self.namespace.as_slice())
                .into_enum()
                .to_string();

            let doc = self.comments_builder.get_comments(&enum_type_path);

            let variants_with_comments = variants_with_comments
                .into_iter()
                .map(|v| Variant { name: v, alias: None })
                .collect();

            self.config
                .enums
                .insert(type_name, Enum { variants: variants_with_comments, doc });
        }
        self
    }

    /// Processes proto message types.
    fn append_msg_type(
        mut self,
        messages: &[DescriptorProto],
        parent_path: &PathBuilder,
        is_nested: bool,
    ) -> Result<Self> {
        for (index, message) in messages.iter().enumerate() {
            let msg_name = message.name();

            let msg_type = GraphQLType::new(msg_name)
                .extend(self.namespace.as_slice())
                .into_object_type();

            if message
                .options
                .as_ref()
                .and_then(|opt| opt.map_entry)
                .unwrap_or_default()
            {
                // map types in protobuf are encoded as nested type
                // https://protobuf.dev/programming-guides/encoding/#maps
                // since we encode it as JSON scalar type in graphQL
                // record that this type is map and ignore it
                self.map_types.insert(msg_type.id());
                continue;
            }

            let msg_path = if is_nested {
                parent_path.extend(PathField::NestedType, index as i32)
            } else {
                parent_path.extend(PathField::MessageType, index as i32)
            };

            // first append the name of current message as namespace
            self.namespace.push(msg_name.to_string());
            self = self.append_enums(&message.enum_type, &PathBuilder::new(&msg_path), true);
            self =
                self.append_msg_type(&message.nested_type, &PathBuilder::new(&msg_path), true)?;
            // then drop it after handling nested types
            self.namespace.pop();

            let mut oneof_fields: Vec<_> = message.oneof_decl.iter().map(|_| Vec::new()).collect();

            let mut ty = Type {
                doc: self.comments_builder.get_comments(&msg_path),
                ..Default::default()
            };

            for (field_index, field) in message.field.iter().enumerate() {
                let field_name = GraphQLType::new(field.name())
                    .extend(self.namespace.as_slice())
                    .into_field();

                let mut cfg_field = Field::default();

                let label = field.label();
                cfg_field.list = matches!(label, Label::Repeated);
                // required only applicable for proto2
                cfg_field.required = matches!(label, Label::Required);

                if let Some(type_name) = &field.type_name {
                    // check that current field is map.
                    // it's done by checking that we've seen this type before
                    // inside the nested type. It works only if we explore nested types
                    // before the current type
                    if self.map_types.contains(&type_name[1..]) {
                        cfg_field.type_of = "JSON".to_string();
                        // drop list option since it is not relevant
                        // when using JSON representation
                        cfg_field.list = false;
                    } else {
                        // for non-primitive types
                        let type_of = graphql_type_from_ref(type_name)?
                            .into_object_type()
                            .to_string();

                        cfg_field.type_of = type_of;
                    }
                } else {
                    let type_of = convert_primitive_type(field.r#type().as_str_name());

                    cfg_field.type_of = type_of;
                }

                let field_path =
                    PathBuilder::new(&msg_path).extend(PathField::Field, field_index as i32);
                cfg_field.doc = self.comments_builder.get_comments(&field_path);

                if let Some(oneof_index) = field.oneof_index {
                    oneof_fields[oneof_index as usize].push((field_name.to_string(), cfg_field));
                } else {
                    ty.fields.insert(field_name.to_string(), cfg_field);
                }
            }

            ty.tag = Some(Tag { id: msg_type.id() });

            if message.oneof_decl.is_empty() {
                self = self.insert_type(msg_type.to_string(), ty);
            } else {
                self = self.insert_oneofs(msg_type.to_string(), ty, oneof_fields);
            }
        }
        Ok(self)
    }

    /// Processes proto service definitions and their methods.
    fn append_query_service(
        mut self,
        services: &[ServiceDescriptorProto],
        parent_path: &PathBuilder,
    ) -> Result<Self> {
        if services.is_empty() {
            return Ok(self);
        }

        for (index, service) in services.iter().enumerate() {
            let service_name = service.name();
            let path = parent_path.extend(PathField::Service, index as i32);

            for (method_index, method) in service.method.iter().enumerate() {
                let field_name = GraphQLType::new(method.name())
                    .extend(self.namespace.as_slice())
                    .push(service_name)
                    .into_method();

                let mut cfg_field = Field::default();
                let mut body = None;

                if let Some(graphql_type) = get_input_type(method.input_type())? {
                    let key = graphql_type.clone().into_field().to_string();
                    let type_of = graphql_type.into_object_type().to_string();
                    let val = Arg {
                        type_of,
                        list: false,
                        required: true,
                        /* Setting it not null by default. There's no way to infer this
                         * from proto file */
                        doc: None,
                        modify: None,
                        default_value: None,
                    };

                    body = Some(Value::String(format!("{{{{.args.{key}}}}}")));
                    cfg_field.args.insert(key, val);
                }

                let output_ty = get_output_type(method.output_type())?
                    .into_object_type()
                    .to_string();
                cfg_field.type_of = output_ty;
                cfg_field.required = true;

                cfg_field.grpc = Some(Grpc {
                    base_url: None,
                    body,
                    batch_key: vec![],
                    headers: vec![],
                    method: field_name.id(),
                });

                let method_path =
                    PathBuilder::new(&path).extend(PathField::Method, method_index as i32);
                cfg_field.doc = self.comments_builder.get_comments(&method_path);

                let ty = self
                    .config
                    .types
                    .entry(self.query.clone())
                    .or_insert_with(|| {
                        self.config.schema.query = Some(self.query.clone());
                        Type::default()
                    });

                ty.fields.insert(field_name.to_string(), cfg_field);
            }
        }
        Ok(self)
    }
}

fn graphql_type_from_ref(name: &str) -> Result<GraphQLType<Unparsed>> {
    if !name.starts_with('.') {
        bail!("Expected fully-qualified name for reference type but got {name}. This is a bug!");
    }

    let name = &name[1..];

    if let Some((package, name)) = name.rsplit_once('.') {
        Ok(GraphQLType::new(name).push(package))
    } else {
        Ok(GraphQLType::new(name))
    }
}

/// Converts proto field types to a custom format.
fn convert_primitive_type(proto_ty: &str) -> String {
    let binding = proto_ty.to_lowercase();
    let proto_ty = binding.strip_prefix("type_").unwrap_or(proto_ty);
    // use Int64Str and Uint64Str to represent 64bit integers as string by default
    // it's how this values are represented in JSON by default in prost
    // see tests in `protobuf::tests::scalars_proto_file`
    match proto_ty {
        "double" | "float" => "Float",
        "int32" | "sint32" | "fixed32" | "sfixed32" => "Int",
        "int64" | "sint64" | "fixed64" | "sfixed64" => "Int64",
        "uint32" => "UInt32",
        "uint64" => "UInt64",
        "bool" => "Boolean",
        "string" => "String",
        "bytes" => "Bytes",
        x => x,
    }
    .to_string()
}

/// Determines the output type for a service method.
fn get_output_type(output_ty: &str) -> Result<GraphQLType<Unparsed>> {
    // type, required
    match output_ty {
        ".google.protobuf.Empty" => {
            // If it's no response is expected, we return an Empty scalar type
            Ok(GraphQLType::new("Empty"))
        }
        _ => {
            // Setting it not null by default. There's no way to infer this from proto file
            graphql_type_from_ref(output_ty)
        }
    }
}

fn get_input_type(input_ty: &str) -> Result<Option<GraphQLType<Unparsed>>> {
    match input_ty {
        ".google.protobuf.Empty" | "" => Ok(None),
        _ => graphql_type_from_ref(input_ty).map(Some),
    }
}

/// The main entry point that builds a Config object from proto descriptor sets.
pub fn from_proto(descriptor_sets: &[FileDescriptorSet], query: &str) -> Result<Config> {
    let mut ctx = Context::new(query);
    for descriptor_set in descriptor_sets.iter() {
        for file_descriptor in descriptor_set.file.iter() {
            ctx.namespace = vec![file_descriptor.package().to_string()];

            if let Some(source_code_info) = &file_descriptor.source_code_info {
                ctx = ctx.with_source_code_info(source_code_info.clone());
            }

            let root_path = PathBuilder::new(&[]);

            ctx = ctx
                .append_enums(&file_descriptor.enum_type, &root_path, false)
                .append_msg_type(&file_descriptor.message_type, &root_path, false)?
                .append_query_service(&file_descriptor.service, &root_path)?;
        }
    }

    let config = AmbiguousType::default()
        .pipe(TreeShake)
        .transform(ctx.config)
        .to_result()?;

    Ok(config)
}

#[cfg(test)]
mod test {
    use anyhow::Result;
    use prost_reflect::prost_types::FileDescriptorSet;
    use tailcall_fixtures::protobuf;

    use super::from_proto;
    use crate::core::config::ConfigModule;

    fn compile_protobuf(files: &[&str]) -> Result<FileDescriptorSet> {
        Ok(protox::compile(files, [protobuf::SELF])?)
    }

    macro_rules! assert_gen {
        ($( $set:expr ), +) => {
            let set = compile_protobuf(&[$( $set ),+]).unwrap();
            let config = from_proto(&[set], "Query").unwrap();
            let config_module = ConfigModule::from(config);
            let result = config_module.to_sdl();
            insta::assert_snapshot!(result);
        };
    }

    #[test]
    fn test_from_proto() {
        // news_enum.proto covers:
        // test for mutation
        // test for empty objects
        // test for optional type
        // test for enum
        // test for repeated fields
        // test for a type used as both input and output
        // test for two types having same name in different packages

        assert_gen!(protobuf::NEWS, protobuf::GREETINGS_A, protobuf::GREETINGS_B);
    }

    #[test]
    fn test_from_proto_no_pkg_file() {
        assert_gen!(protobuf::NEWS_NO_PKG);
    }

    #[test]
    fn test_from_proto_no_service_file() {
        assert_gen!(protobuf::NEWS_NO_SERVICE);
    }

    #[test]
    fn test_greetings_proto_file() {
        assert_gen!(protobuf::GREETINGS, protobuf::GREETINGS_MESSAGE);
    }

    #[test]
    fn test_config_from_sdl() -> Result<()> {
        let set =
            compile_protobuf(&[protobuf::NEWS, protobuf::GREETINGS_A, protobuf::GREETINGS_B])?;

        let set1 = compile_protobuf(&[protobuf::NEWS])?;
        let set2 = compile_protobuf(&[protobuf::GREETINGS_A])?;
        let set3 = compile_protobuf(&[protobuf::GREETINGS_B])?;

        let actual = from_proto(&[set.clone()], "Query")?.to_sdl();
        let expected = from_proto(&[set1, set2, set3], "Query")?.to_sdl();

        pretty_assertions::assert_eq!(actual, expected);
        Ok(())
    }

    #[test]
    fn test_required_types() {
        // required fields are deprecated in proto3 (https://protobuf.dev/programming-guides/dos-donts/#add-required)
        // this example uses proto2 to test the same.
        // for proto3 it's guaranteed to have a default value (https://protobuf.dev/programming-guides/proto3/#default)
        // and our implementation (https://github.com/tailcallhq/tailcall/pull/1537) supports default values by default.
        // so we do not need to explicitly mark fields as required.

        assert_gen!(protobuf::PERSON);
    }

    #[test]
    fn test_movies() {
        assert_gen!(protobuf::MOVIES);
    }

    #[test]
    fn test_nested_types() {
        assert_gen!(protobuf::NESTED_TYPES);
    }

    #[test]
    fn test_map_types() {
        assert_gen!(protobuf::MAP);
    }

    #[test]
    fn test_optional_fields() {
        assert_gen!(protobuf::OPTIONAL);
    }

    #[test]
    fn test_scalar_types() {
        assert_gen!(protobuf::SCALARS);
    }

    #[test]
    fn test_oneof_types() {
        assert_gen!(protobuf::ONEOF);
    }
}
