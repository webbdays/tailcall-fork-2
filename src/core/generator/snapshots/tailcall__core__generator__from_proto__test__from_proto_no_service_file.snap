---
source: src/core/generator/from_proto.rs
expression: result
---

